File: sdk/samples/WPFSamples/CustomPopupTextBox/csharp/editablepopup.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Controls.Primitives;
using System.Windows.Data;
using System.Windows.Documents;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using System.Windows.Shapes;

namespace CustomPopupTextBox
{
    /// <summary>
    /// This control is a TextBox that acts like a Popup.  To get a control
    /// to behave like a Popup, you should define the IsOpen, Placement, 
    /// PlacementRectangle, PlacementTarget, HorizontalOffset, and
    /// VerticalOffset properties and then call Popup.CreateRoot to
    /// bind these six properties to their counterparts on the 
    /// internal Popup.
    /// </summary>
    public partial  class EditablePopup : System.Windows.Controls.TextBox
    {
        Popup _parentPopup;

        public EditablePopup()
            : base()
        {
        }

        static EditablePopup()
        {
            //This OverrideMetadata call tells the system that this element 
            //wants to provide a style that is different than its base class.
            //This style is defined in themes\generic.xaml
            DefaultStyleKeyProperty.OverrideMetadata(typeof(EditablePopup), 
                new FrameworkPropertyMetadata(typeof(EditablePopup)));
        }

        #region Properties to implement Popup Behavior

        //Placement
        public static readonly DependencyProperty PlacementProperty =
                    Popup.PlacementProperty.AddOwner(typeof(EditablePopup));

        public PlacementMode Placement
        {
            get { return (PlacementMode)GetValue(PlacementProperty); }
            set { SetValue(PlacementProperty, value); }
        }

        //PlacementTarget
        public static readonly DependencyProperty PlacementTargetProperty =
           Popup.PlacementTargetProperty.AddOwner(typeof(EditablePopup));

        public UIElement PlacementTarget
        {
            get { return (UIElement)GetValue(PlacementTargetProperty); }
            set { SetValue(PlacementTargetProperty, value); }
        }

        //PlacementRectangle
        public static readonly DependencyProperty PlacementRectangleProperty =
                    Popup.PlacementRectangleProperty.AddOwner(typeof(EditablePopup));

        public Rect PlacementRectangle
        {
            get { return (Rect)GetValue(PlacementRectangleProperty); }
            set { SetValue(PlacementRectangleProperty, value); }
        }

        //HorizontalOffset
        public static readonly DependencyProperty HorizontalOffsetProperty =
            Popup.HorizontalOffsetProperty.AddOwner(typeof(EditablePopup));

        public new double HorizontalOffset
        {
            get { return (double)GetValue(HorizontalOffsetProperty); }
            set { SetValue(HorizontalOffsetProperty, value); }
        }

        //VerticalOffset
        public static readonly DependencyProperty VerticalOffsetProperty =
                Popup.VerticalOffsetProperty.AddOwner(typeof(EditablePopup));

        public new double VerticalOffset
        {
            get { return (double)GetValue(VerticalOffsetProperty);  }
            set { SetValue(VerticalOffsetProperty, value); }
        }

        
        public static readonly DependencyProperty IsOpenProperty =
                Popup.IsOpenProperty.AddOwner(
                        typeof(EditablePopup),
                        new FrameworkPropertyMetadata(
                                false,
                                FrameworkPropertyMetadataOptions.BindsTwoWayByDefault,
                                new PropertyChangedCallback(OnIsOpenChanged)));

        public bool IsOpen
        {
            get { return (bool)GetValue(IsOpenProperty); }
            set { SetValue(IsOpenProperty, value); }
        }

        private static void OnIsOpenChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            EditablePopup ctrl = (EditablePopup)d;

            if ((bool)e.NewValue)
            {
                if (ctrl._parentPopup == null)
                {
                    ctrl.HookupParentPopup();
                }
            }
        }

        //Create the Popup and attach the CustomControl to it.
        private void HookupParentPopup()
        {

            _parentPopup = new Popup();

            _parentPopup.AllowsTransparency = true;

            Popup.CreateRootPopup(_parentPopup, this);
        }

        #endregion

    }
}

File: sdk/samples/WPFSamples/CustomPopupTextBox/visualbasic/editablepopup.vb

Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Controls.Primitives
Imports System.Windows.Data
Imports System.Windows.Documents
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Imaging
Imports System.Windows.Shapes


' <summary>
' This control is a TextBox that acts like a Popup.  To get a control
' to behave like a Popup, you should define the IsOpen, Placement, 
' PlacementRectangle, PlacementTarget, HorizontalOffset, and
' VerticalOffset properties and then call Popup.CreateRoot to
' bind these six properties to their counterparts on the 
' internal Popup.
' </summary>

Partial Class EditablePopup
    Inherits System.Windows.Controls.TextBox '

    Private _parentPopup As Popup

    Public Sub New()

    End Sub 'New


    Shared Sub New()
        'This OverrideMetadata call tells the system that this element 
        'wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(EditablePopup), _
                        New FrameworkPropertyMetadata(GetType(EditablePopup)))

    End Sub 'New

#Region "Properties to implement Popup Behavior"

    'Placement
    Public Shared PlacementProperty As DependencyProperty = _
                        Popup.PlacementProperty.AddOwner(GetType(EditablePopup))


    Public Property Placement() As PlacementMode
        Get
            Return CType(GetValue(PlacementProperty), PlacementMode)
        End Get
        Set(ByVal value As PlacementMode)
            SetValue(PlacementProperty, Value)
        End Set
    End Property

    'PlacementTarget
    Public Shared PlacementTargetProperty As DependencyProperty = _
                        Popup.PlacementTargetProperty.AddOwner(GetType(EditablePopup))


    Public Property PlacementTarget() As UIElement
        Get
            Return CType(GetValue(PlacementTargetProperty), UIElement)
        End Get
        Set(ByVal value As UIElement)
            SetValue(PlacementTargetProperty, Value)
        End Set
    End Property


    'PlacementRectangle
    Public Shared PlacementRectangleProperty As DependencyProperty = _
                        Popup.PlacementRectangleProperty.AddOwner(GetType(EditablePopup))


    Public Property PlacementRectangle() As Rect
        Get
            Return CType(GetValue(PlacementRectangleProperty), Rect)
        End Get
        Set(ByVal value As Rect)
            SetValue(PlacementRectangleProperty, Value)
        End Set
    End Property

    'HorizontalOffset
    Public Shared HorizontalOffsetProperty As DependencyProperty = _
                        Popup.HorizontalOffsetProperty.AddOwner(GetType(EditablePopup))


    Public Overloads Property HorizontalOffset() As Double
        Get
            Return System.Convert.ToDouble(GetValue(HorizontalOffsetProperty))
        End Get
        Set(ByVal value As Double)
            SetValue(HorizontalOffsetProperty, value)
        End Set
    End Property

    'VerticalOffset
    Public Shared VerticalOffsetProperty As DependencyProperty = _
                        Popup.VerticalOffsetProperty.AddOwner(GetType(EditablePopup))

    Public Overloads Property VerticalOffset() As Double
        Get
            Return System.Convert.ToDouble(GetValue(VerticalOffsetProperty))
        End Get
        Set(ByVal value As Double)
            SetValue(VerticalOffsetProperty, value)
        End Set
    End Property

    Public Shared IsOpenProperty As DependencyProperty = _
                Popup.IsOpenProperty.AddOwner(GetType(EditablePopup), _
                                     New FrameworkPropertyMetadata(False, _
                                     FrameworkPropertyMetadataOptions.BindsTwoWayByDefault, _
                                     New PropertyChangedCallback(AddressOf OnIsOpenChanged)))


    Public Property IsOpen() As Boolean
        Get
            Return CBool(GetValue(IsOpenProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(IsOpenProperty, value)
        End Set
    End Property

    Private Shared Sub OnIsOpenChanged(ByVal d As DependencyObject, _
                                       ByVal e As DependencyPropertyChangedEventArgs)

        Dim ctrl As EditablePopup = CType(d, EditablePopup)

        If CBool(e.NewValue) Then
            If ctrl._parentPopup Is Nothing Then
                ctrl.HookupParentPopup()
            End If
        End If

    End Sub 'OnIsOpenChanged


    'Create the Popup and attach the CustomControl to it.
    Private Sub HookupParentPopup()

        _parentPopup = New Popup()

        _parentPopup.AllowsTransparency = True

        Popup.CreateRootPopup(_parentPopup, Me)

    End Sub 'HookupParentPopup

#End Region
End Class 'EditablePopup 
